### PR TITLE
initial commit for DNS local resolver

### PR DESCRIPTION
**Pull request by jaytaph**
This is the initial commit for the DNS resolver component. 
In consists of a single entrypoint which returns a DNSEntry. It is up to the caller to actually extract the correct ips it wants (either the ipv4() or ipv6().

There are 3 resolvers that are called in sequence until one of them returns a hit:

  - caching resolver
  - local table resolver
  - remote resolver

The caching resolver has a fixed size cache that will retrieve quickly a dns record if it is available.
The local table resolver contains a list of items that will be checked against (including wildcards). This allows to override hosts just like an /etc/host file or dnsmasq would do
The remote resolver calls hickory-dns, which will resolve the query.

It is not 100% finished, but the major work is done. As soon as we have the config store in place, we can use that for configuration purposes.

Another things left to do is expiring of DNS records.

```
    Finished test [unoptimized + debuginfo] target(s) in 0.08s
     Running unittests src/lib.rs (target/debug/deps/gosub_engine-761d2bb186cfbf97)
2023-11-17T21:57:37.971Z INFO  [gosub_engine::dns] Resolving example.org for Ipv4
2023-11-17T21:57:37.971Z DEBUG [gosub_engine::dns] Trying resolver: cache resolver
2023-11-17T21:57:37.971Z DEBUG [gosub_engine::dns] Trying resolver: local table resolver
2023-11-17T21:57:37.971Z TRACE [gosub_engine::dns::local] example.org: not found in local table
2023-11-17T21:57:37.971Z DEBUG [gosub_engine::dns] Trying resolver: remote resolver
2023-11-17T21:57:37.971Z TRACE [gosub_engine::dns::remote] example.org: resolving with ["ipv4"]
2023-11-17T21:57:37.977Z TRACE [gosub_engine::dns::remote] example.org: found ipv4 address 93.184.216.34
2023-11-17T21:57:37.978Z DEBUG [gosub_engine::dns] Found entry DnsEntry { domain: "example.org", ips: [93.184.216.34], has_ipv4: true, has_ipv6: false, iter: 0, expires: 0 }
2023-11-17T21:57:37.978Z TRACE [gosub_engine::dns::cache] example.org: announcing to cache
2023-11-17T21:57:37.978Z TRACE [gosub_engine::dns::cache] adding new entry to cache
ipv4: 93.184.216.34
Took 6626 microseconds.
2023-11-17T21:57:37.978Z INFO  [gosub_engine::dns] Resolving example.org for Ipv6
2023-11-17T21:57:37.978Z DEBUG [gosub_engine::dns] Trying resolver: cache resolver
2023-11-17T21:57:37.978Z TRACE [gosub_engine::dns::cache] example.org: no ipv6 addresses found in entry
2023-11-17T21:57:37.978Z DEBUG [gosub_engine::dns] Trying resolver: local table resolver
2023-11-17T21:57:37.978Z TRACE [gosub_engine::dns::local] example.org: not found in local table
2023-11-17T21:57:37.978Z DEBUG [gosub_engine::dns] Trying resolver: remote resolver
2023-11-17T21:57:37.978Z TRACE [gosub_engine::dns::remote] example.org: resolving with ["ipv6"]
2023-11-17T21:57:37.986Z TRACE [gosub_engine::dns::remote] example.org: found ipv6 address 2606:2800:220:1:248:1893:25c8:1946
2023-11-17T21:57:37.986Z DEBUG [gosub_engine::dns] Found entry DnsEntry { domain: "example.org", ips: [2606:2800:220:1:248:1893:25c8:1946], has_ipv4: false, has_ipv6: true, iter: 0, expires: 0 }
2023-11-17T21:57:37.986Z TRACE [gosub_engine::dns::cache] example.org: announcing to cache
2023-11-17T21:57:37.986Z TRACE [gosub_engine::dns::cache] example.org: updating existing entry to cache
2023-11-17T21:57:37.986Z TRACE [gosub_engine::dns::cache] current entries: [93.184.216.34]
2023-11-17T21:57:37.986Z TRACE [gosub_engine::dns::cache] new entries: [2606:2800:220:1:248:1893:25c8:1946]
2023-11-17T21:57:37.986Z TRACE [gosub_engine::dns::cache] merged entries: [93.184.216.34, 2606:2800:220:1:248:1893:25c8:1946]
ipv6: 2606:2800:220:1:248:1893:25c8:1946
Took 8253 microseconds.
2023-11-17T21:57:37.986Z INFO  [gosub_engine::dns] Resolving example.org for Ipv4
2023-11-17T21:57:37.986Z DEBUG [gosub_engine::dns] Trying resolver: cache resolver
2023-11-17T21:57:37.986Z TRACE [gosub_engine::dns::cache] example.org: found in cache with correct resolve type
2023-11-17T21:57:37.986Z DEBUG [gosub_engine::dns] Found entry DnsEntry { domain: "example.org", ips: [93.184.216.34, 2606:2800:220:1:248:1893:25c8:1946], has_ipv4: true, has_ipv6: true, iter: 0, expires: 0 }
2023-11-17T21:57:37.986Z TRACE [gosub_engine::dns::cache] example.org: announcing to cache
2023-11-17T21:57:37.986Z TRACE [gosub_engine::dns::cache] example.org: updating existing entry to cache
2023-11-17T21:57:37.986Z TRACE [gosub_engine::dns::cache] current entries: [93.184.216.34, 2606:2800:220:1:248:1893:25c8:1946]
2023-11-17T21:57:37.986Z TRACE [gosub_engine::dns::cache] new entries: [93.184.216.34, 2606:2800:220:1:248:1893:25c8:1946]
2023-11-17T21:57:37.986Z TRACE [gosub_engine::dns::cache] merged entries: [93.184.216.34, 2606:2800:220:1:248:1893:25c8:1946]
ipv4: 93.184.216.34
Took 57 microseconds.
2023-11-17T21:57:37.986Z INFO  [gosub_engine::dns] Resolving example.org for Both
2023-11-17T21:57:37.986Z DEBUG [gosub_engine::dns] Trying resolver: cache resolver
2023-11-17T21:57:37.986Z TRACE [gosub_engine::dns::cache] example.org: found in cache with correct resolve type
2023-11-17T21:57:37.986Z DEBUG [gosub_engine::dns] Found entry DnsEntry { domain: "example.org", ips: [93.184.216.34, 2606:2800:220:1:248:1893:25c8:1946], has_ipv4: true, has_ipv6: true, iter: 0, expires: 0 }
2023-11-17T21:57:37.986Z TRACE [gosub_engine::dns::cache] example.org: announcing to cache
2023-11-17T21:57:37.986Z TRACE [gosub_engine::dns::cache] example.org: updating existing entry to cache
2023-11-17T21:57:37.986Z TRACE [gosub_engine::dns::cache] current entries: [93.184.216.34, 2606:2800:220:1:248:1893:25c8:1946]
2023-11-17T21:57:37.986Z TRACE [gosub_engine::dns::cache] new entries: [93.184.216.34, 2606:2800:220:1:248:1893:25c8:1946]
2023-11-17T21:57:37.986Z TRACE [gosub_engine::dns::cache] merged entries: [93.184.216.34, 2606:2800:220:1:248:1893:25c8:1946]
ipv4: 93.184.216.34
ipv6: 2606:2800:220:1:248:1893:25c8:1946
Took 45 microseconds.
```

### CODE REVIEW

File: src/dns.rs
Comment (jaytaph) [inline src/dns.rs:15]: This needs to be filled in later.. this pr just does the local resolver

File: src/dns/local.rs
Comment (jaytaph) [inline src/dns/local.rs:19]: if a domain has multiple ips, we store a round robin pointer so the next call to resolve will return the next ip in the list.

File: src/dns/local.rs
Comment (Sharktheone) [inline src/dns/local.rs:39]: Here I would change it then (when we us a vec in DNSEntry)

```suggestion
            if entry.domain_name == domain {
                ips.append(entry.ips);
                break;
            } else if WildMatch::new(&entry.domain_name).matches(domain) { //when we have a more specific domain name don't use the WildMatch
                 wildcard_ips.append(entry.ips);
                 break;
            }

```

**Review by Sharktheone — CHANGES_REQUESTED** [review]

File: src/dns.rs
Comment (jaytaph) [reply src/dns.rs:15]: yes, this will be behind a trait

File: src/dns.rs
Comment (jaytaph) [reply src/dns.rs:12]: The idea is that the resolver always returns a single address (either ipv4 or v6), depending on settings (not sure if these are resolver settings, or user agent settings). If there are multiple ips for a single host, we roundrobin them and this should be done by the resolver.

It is possible to change this into a vec if we find multiple entries, but we should have our round-robin pointer in a dns entry struct as well.

File: src/dns.rs
Comment (Sharktheone) [inline src/dns.rs:55]: I think, the expect shouldn't be here. This error is recoverable (just ignore the ip ptr), so maybe just use a Result or an Option / "silently" recover the error may be better.

File: src/dns/local.rs
Comment (Sharktheone) [inline src/dns/local.rs:42]: Maybe put this over the wildcard match because it can optimize the time this a tiny bit. (But feel free to ignore)

File: src/dns/local.rs
Comment (Sharktheone) [inline src/dns/local.rs:57]: Here again, don't use expect, I mean this function already returns a Result. Maybe it would be nice to implement the functionality for the `?` syntax. If this is not already done, i can do this if you want.

**Review by Sharktheone — CHANGES_REQUESTED** [review]

File: src/dns.rs
Comment (jaytaph) [reply src/dns.rs]: The DNS resolver is now a trait.. I will have a few: the global dns resolver (that uses hickory dns), the local override table resolver and maybe others in the future. The dns components will consist of a global dns resolver, a local dns resolver, and a cache layer.  First, the resolver checks the cache layer. If time expired for the entry, or the entry is not found, the local dns resolver is queried. If there is nothing found, it will check the global dns resolver.

File: src/dns.rs
Comment (jaytaph) [reply src/dns.rs]: Changed the dns entry as a whole now

File: src/dns.rs
Comment (jaytaph) [inline src/dns.rs:13]: as stated in the todo: i'm not sure if both expired and last-used should be part of the ip entry. I think the only place where ttl / expiration is important, would be the caching layer. This means the resolver should return a tuple (ipaddress, ttl) so the layer above can cache it accordingly. Time last used may not be needed.

File: src/dns.rs
Comment (jaytaph) [inline src/dns.rs:25]: The DNS entry now consists of a domain name, and 2 lists of ip addresses, the ipv4 andipv6. It now contains the roundrobin pointers as well. The whole round robin is encapsulated in the DNSEntry so we can simply do next_ipv4() and not need to worry about it from the outside world.

File: src/dns.rs
Comment (jaytaph) [inline src/dns.rs:84]: I don't know exactly who will be in charge of this, but somewhere we must decide if we resolve the ipv4 or ipv6 of a domain. This is the ResolveType we can add to the resolve() function.

File: src/dns/cache.rs
Comment (jaytaph) [inline src/dns/cache.rs:4]: The cache has a fixed size of max_entries elements and uses a lru strategy for evicting old dns entries. It's not yet complete implemented, but the lru seems to work according to the tests

File: src/dns/local.rs
Comment (jaytaph) [inline src/dns/local.rs:14]: The wildcard matcher has been replaced by the tree lookup crate. It will match deeper wildcards better (*.example.org would match foo.bar.baz.example.org, and we don't want that).

Also, we now do not have to do a difficult process of matching both specific and wildcard matching.

File: src/dns/local.rs
Comment (jaytaph) [inline src/dns/local.rs:54]: Could be configurable by  "dns.prefer.ipv6.over.ipv4" maybe

File: src/dns/local.rs
Comment (jaytaph) [inline src/dns/local.rs:137]: we can remove this func

File: src/types.rs
Comment (jaytaph) [inline src/types.rs:50]: I like the way we can simply create errors and return them in Result<> , but should we add them all inside a generic enum like this? Wouldn't we want the errors to be per component? (`src/dns/errors.rs`, `src/html5/errors.rs` etc)

File: src/dns.rs
Comment (emwalker) [inline src/dns.rs:35]: Should this be an enum?  Also, curious why we aren't making use of a simple wrapper around an existing Rust library.  This looks more low-level than I was anticipating.

Comment (jaytaph) [reply src/dns.rs:35]: We are using hickory dns in the remote resolver. I'll draw up a diagram to give you an idea of the architecture (its not too complicated)

File: src/dns.rs
Comment (jaytaph) [inline src/dns.rs:24]: an entry has either resolved ipv4 addresses, resolver ipv6 addresses or both. WHen we have a record in cache, but we want to have ipv6 addresses for that record, we can see if these are already resolved for this record.

File: src/dns.rs
Comment (jaytaph) [inline src/dns.rs:31]: We don't use the expires time yet. This will be taken from a TTL that is returned by the remote resolver.

File: src/dns/cache.rs
Comment (jaytaph) [inline src/dns/cache.rs:40]: Each resolver has an "announce", which will be called when a new dns entry has been found. It is only used by the cache resolver to store this in cache.

File: src/types.rs
Comment (Kiyoshika) [reply src/types.rs:50]: I think it'd make sense to make them component specific now that the list is growing (and will grow more in the future.) I remember this was added maybe a month or two back when there were only 2-3 error types, but now I'd be in favour to move groups of them into their own components

**Review by Kiyoshika — APPROVED** [review]

File: src/dns.rs
Comment (emwalker) [inline src/dns.rs:82]: This cleanup was confusing for me, and now I see what the source of the confusion is.  Iterators should be implemented as separate structs that fall out of scope / are discarded after they're exhausted.  So a cleanup / reset step shouldn't be needed. See `TreeIterator` for an example of this pattern.  If there's a cleanup step, it suggests a new struct needs to be added.  

But in this case, a separate struct for the iterator isn't needed, because you can get an iterator over the items you want through other means:

```patch
diff --git a/src/dns.rs b/src/dns.rs
index 36b6a6d..cb2b42e 100644
--- a/src/dns.rs
+++ b/src/dns.rs
@@ -70,24 +70,9 @@ impl DnsEntry {
     fn ipv6(&self) -> Vec<IpAddr> {
         self.ips.iter().filter(|x| x.is_ipv6()).copied().collect()
     }
-}
-
-impl Iterator for DnsEntry {
-    type Item = IpAddr;
-
-    /// With next() you can simply iterate over all the ips in the list
-    fn next(&mut self) -> Option<Self::Item> {
-        if self.iter >= self.ips.len() {
-            // reset iterator at the end
-            self.iter = 0;
-
-            return None;
-        }
-
-        let ip = self.ips[self.iter];
-        self.iter += 1;
 
-        Some(ip)
+    fn iter(&self) -> impl Iterator<Item = &IpAddr> {
+        self.ips.iter()
     }
 }
```

File: src/dns.rs
Comment (emwalker) [inline src/dns.rs:119]: Dead code.

File: src/dns.rs
Comment (emwalker) [inline src/dns.rs:158]: Nice use of the log crate.

File: src/dns/remote.rs
Comment (emwalker) [inline src/dns/remote.rs:31]: It strikes me as unusual that we're using a static string here as a type marker.

File: src/types.rs
Comment (emwalker) [reply src/types.rs:50]: I'm guessing the rule of thumb on when to create a new error type and when to stick with an existing type is when you want to match against a specific kind of error and not other errors.  So the number of error types becomes the number of specific kinds of error you need to match on in high-level application code.

The matching in error handling code is what is important here.  I can imagine needing a total of three kinds of error in all, or 20, depending on what kind of error handling is needed.

**Review by emwalker — APPROVED** [review]
It would be nice if you would update the iterator code I left a note on.  But I also don't want to hold up this PR.  I suspect some of the details will change later.

This code makes me nervous, because the network stack exposes a large surface area for security attacks.  But I can also see why it's needed.  I do wonder whether there's already some canonical Rust crates out there that do some or all of this.  In the Rust ecosystem, semi-official Rust crates (e.g., `serde`) seem to play a bigger role than third-party libraries in other language ecosystems.

File: src/dns/remote.rs
Comment (Sharktheone) [reply src/dns/remote.rs:31]: I guess we can just use the ResolverType itself and if it is Both we Push `ResolverType::Ipv4` and `ReolverType::Ipv6`

File: src/dns.rs
Comment (Sharktheone) [inline src/dns.rs:109]: Maybe add a default implementation here, so we only need to override it in the CacheResolver
```suggestion
    fn announce(&mut self, domain: &str, entry: &DnsEntry) {};
```

**Review by Sharktheone — APPROVED** [review]
Looks fine to me, just some simple things. But you can ignore them if you want.

File: src/dns.rs
Comment (jaytaph) [reply src/dns.rs:109]: ah, didn't know this was possible.. very nice

File: src/dns.rs
Comment (jaytaph) [reply src/dns.rs:134]: I would not worry about this too much. It's not used in any thight loops, and the first push will do an allocation for 4 elements anyway. 
So instead of with_capacity(3), you get a with_capacity(4), which is ok enough.

File: src/dns.rs
Comment (jaytaph) [reply src/dns.rs:87]: 👍

File: src/dns/remote.rs
Comment (jaytaph) [reply src/dns/remote.rs:31]: 👍

File: src/dns.rs
Comment (jaytaph) [reply src/dns.rs:119]: removed

**Review by Sharktheone — APPROVED** [review]

File: src/dns.rs
Comment (emwalker) [inline src/dns.rs:76]: After I thought about it more, I realized you might have been wanting to use the iterator trait as a state machine for implementing round-robin.  If so, I think something other than `trait Iterator` would be used for this, e.g., a custom method that updates the state to the next ip address, possibly in a new struct that this struct delegates to.

Comment (jaytaph) [reply src/dns.rs:76]: i've actually removed the round robin inside the dns entries because it ended up a bit too complex. I leave it up to the caller to deal with round robin. There might be other ways they want to use the set of resolved ips: when one fails, use another, etc..

**Pull request merged**
